src/predict_db.cc

```diff
@@ -83,67 +82,91 @@ bool PredictDb::Build(const predict::RawData& data) {
   // create predict db
   int data_size = data.size();
   const size_t kReservedSize = 1024;
-  const size_t kCandidatesPerKey = 10;
-  const size_t kKeyStringSize = 10;
-  const size_t kCandidateStringSize = 10;
-  size_t estimated_size = kReservedSize + data_size * (
-      sizeof(Array<table::Entry>) + kKeyStringSize +
-      kCandidatesPerKey * (sizeof(table::Entry) + kCandidateStringSize));
+
+  size_t entry_count = 0;
+  for (const auto& kv : data) {
+    entry_count += kv.second.size();
+  }
+  StringTableBuilder string_table;
+  vector<table::Entry> entries(entry_count);
+  vector<const char*> keys;
+  vector<int> values(data_size);
+  keys.reserve(data_size);
+  int i = 0;
+  for (const auto& kv : data) {
+    if (kv.second.empty())
+      continue;
+    for (const auto& candidate : kv.second) {
+      string_table.Add(candidate.text, candidate.weight,
+                       &entries[i].text.str_id());
+      entries[i].weight = float(candidate.weight);
+      ++i;
+    }
+    keys.push_back(kv.first.c_str());
+  }
+  size_t key_trie_array_size;
+  size_t key_trie_image_size;
+  {
+    // build a temporary key_trie to get size
+    Darts::DoubleArray key_trie;
+    key_trie.build(data_size, &keys[0], NULL, &values[0]);
+    key_trie_array_size = key_trie.size();
+    key_trie_image_size = key_trie.total_size();
+    values.clear();
+  }
+  // this writes to entry vector, which should be copied to entry array later
+  string_table.Build();
+  size_t value_trie_image_size = string_table.BinarySize();
+  size_t array_size =
+      data_size * (sizeof(Array<table::Entry>) - sizeof(table::Entry)) +
+      entry_count * sizeof(table::Entry);
+  size_t estimated_size =
+      kReservedSize + array_size + key_trie_image_size + value_trie_image_size;
   if (!Create(estimated_size)) {
     LOG(ERROR) << "Error creating predict db file '" << file_name() << "'.";
     return false;
   }
   // create metadata in the beginning of file
-  auto metadata = Allocate<predict::Metadata>();
-  if (!metadata) {
+  metadata_ = Allocate<predict::Metadata>();
+  if (!metadata_) {
     LOG(ERROR) << "Error creating metadata in file '" << file_name() << "'.";
     return false;
   }
-  // build trie
-  StringTableBuilder string_table;
-  int entry_count = 0;
-  vector<const char*> keys;
-  vector<int> values;
-  keys.reserve(data_size);
-  values.reserve(data_size);
+
+  // copy from entry vector to entry array
+  table::Entry* entry = &entries[0];
```

Comment:
a better name for the pointer can be `available_entries`.

Reply:
a pointer type is the same as an array.
in this particular case, it's not just pointing to one entry but the start address of an array of entries.
so we can use plural in the name.